Web Inspector: Pull CSS property keyword completions from the backend instead of hardcoding them

stwrt opened the pull request:
#### 68e73e36fe065e4229d37cf0c686e5bd7a64bf53
<pre>
Web Inspector: Pull CSS property keyword completions from the backend instead of hardcoding them
<a href="https://bugs.webkit.org/show_bug.cgi?id=303627">https://bugs.webkit.org/show_bug.cgi?id=303627</a>
<a href="https://rdar.apple.com/problem/165914089">rdar://problem/165914089</a>

Reviewed by Tim Nguyen and Devin Rousso.

The Web Inspector&apos;s CSS autocomplete system previously hardcoded all valid
keyword values for CSS properties in CSSKeywordCompletions.js. This meant
that when new CSS values were added to WebCore, they had to be manually
duplicated in the Inspector frontend, and there was no way to filter values
based on runtime settings (like feature flags).

This change extends getSupportedCSSProperties() to return the valid keyword
values for each property, pulled directly from CSSProperties.json. The
frontend now uses these backend-provided values instead of the hardcoded
ones, ensuring completions stay in sync with WebCore and respect feature
flags.

Key changes:
- Added validKeywordsForProperty() to return keyword values from CSSProperties.json
- Added isKeywordValidForPropertyValues() to filter keywords by settings flags
- Added CSSParserContext(const Settings&amp;) constructor for easy context creation
- Updated getSupportedCSSProperties() to include filtered values in the response
- Added settings-flag to grid-lanes/inline-grid-lanes values in CSSProperties.json

* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/scripts/process-css-properties.py:
(GenerateCSSPropertyNames._generate_css_property_names_gperf_prelude):
(GenerateCSSPropertyNames):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getSupportedCSSProperties):

Canonical link: <a href="https://commits.webkit.org/304195@main">https://commits.webkit.org/304195@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e6cc5e46a546687c8f8ebd4aa37f9eccbf91648

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134862 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46107 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142374 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/55b86ebf-e71c-4d15-a286-a119aabde4c7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136731 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7905 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7137 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c87ab974-8edb-4035-96e5-d8dca04b90c7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137808 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120880 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83891 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f73ed9b3-8a8f-4340-928f-1861c86f3d7e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5412 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3028 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2967 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145072 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6965 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39608 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7024 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5844 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111777 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28356 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5251 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117160 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60895 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7012 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35335 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6796 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70590 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Checked out pull request; Found 49824 issues") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7032 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6906 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
I don't think this takes in account removing values gated behind feature flags

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -927,6 +927,16 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::CSS::C
             if (values->length())
                 property->setValues(WTFMove(values));
+        } else {
+            // For properties that aren't keyword-fast-path eligible (e.g., display),
```
can we add maybe a test case for something like `display` in `LayoutTests/inspector/css/getSupportedCSSProperties.html`?

stwrt replied:
Will add a test.

dcrousso reviewed (approved):
neat!

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:
```diff
@@ -1227,9 +1227,6 @@ WI.CSSKeywordCompletions._propertyKeywordMap = {
     "direction": [
         "ltr", "rtl",
     ],
-    "display": [
```
let's keep this for now as your new code will not exist when inspecting an older version of iOS, so we'd still wanna have some completions for that

stwrt replied:
I'll add this back in. This was removed while I was doing testing.

nt1m reviewed (changes requested):
See my comment above, values that are behind a disabled by default feature flags will be pulled with this PR

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -121,6 +121,45 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
 {
 }
 
+CSSParserContext::CSSParserContext(const Settings& settings)
```
Can we make the other constructor re-use this one to avoid duplication/extra boilerplate?

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -78,46 +78,52 @@ CSSParserContext::CSSParserContext(const Document& document)
-    , popoverAttributeEnabled { document.settings().popoverAttributeEnabled() }
-    , sidewaysWritingModesEnabled { document.settings().sidewaysWritingModesEnabled() }
-    , cssTextWrapPrettyEnabled { document.settings().cssTextWrapPrettyEnabled() }
-    , thumbAndTrackPseudoElementsEnabled { document.settings().thumbAndTrackPseudoElementsEnabled() }
```
nit: Can you restore the indentation as it was?

stwrt replied:
The style checker was demanding this, but I can change it back.

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -78,46 +78,52 @@ CSSParserContext::CSSParserContext(const Document& document)
-    , springTimingFunctionEnabled { document.settings().springTimingFunctionEnabled() }
+    : CSSParserContext(document.settings())
+{
+    this->baseURL = sheetBaseURL.isNull() ? document.baseURL() : sheetBaseURL;
```
```suggestion
    baseURL = sheetBaseURL.isNull() ? document.baseURL() : sheetBaseURL;
```

nt1m commented on Source/WebCore/css/scripts/process-css-properties.py:
```diff
@@ -3359,6 +3362,138 @@ def _generate_css_property_id_text_stream(self, *, to):
+        for prop in self.properties_and_descriptors.style_properties.all:
+            if hasattr(prop, 'values') and prop.values:
+                # Filter to only include values that have a valid keyword_term (actual keywords)
+                keyword_values = [v for v in prop.values if hasattr(v, 'value_keyword_name') and v.value_keyword_name]
```
nit: here and in the rest function, we don't usually use abbreviations like `v` in WebKit.

nt1m commented on Source/WebCore/css/scripts/process-css-properties.py:
@weinig has tests for the auto-generated parts of this script, maybe it would be worth extending them?

nt1m commented on Source/WebCore/css/scripts/process-css-properties.py:
```diff
@@ -2967,7 +2967,10 @@ def _generate_css_property_names_gperf_prelude(self, *, to):
+                "CSSParserContext.h",
                 "CSSProperty.h",
+                "CSSValueKeywords.h",
+                "DeprecatedGlobalSettings.h",
```
Do we really need `DeprecatedGlobalSettings` included in the auto-generated files?

stwrt replied:
Yes, we need to include this header.

```
CSSPropertyNames.gperf:8047:16: error: use of undeclared identifier 'DeprecatedGlobalSettings'
 8047 |         return DeprecatedGlobalSettings::attachmentElementEnabled();
1 error generated.
```

nt1m commented on Source/WebCore/css/scripts/process-css-properties.py:
```diff
@@ -2967,7 +2967,10 @@ def _generate_css_property_names_gperf_prelude(self, *, to):
             to=to,
             headers=[
                 "BoxSides.h",
+                "CSSParserContext.h",
```
Do we really need `CSSParserContext` included in the auto-generated files?

stwrt replied:
Yes, we need to include this header, because we appear to be accessing internal member sof the class.

```
return context.sidewaysWritingModesEnabled;
```

nt1m reviewed (approved):
seems ok, haven't reviewed too precisely the non CSS bits, but I trust @dcrousso for the other parts

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:
```diff
@@ -231,8 +231,10 @@ WI.CSSKeywordCompletions.addCustomCompletions = function(properties)
-            WI.CSSKeywordCompletions.addPropertyCompletionValues(property.name, property.values);
+        if (property.values) {
+            // Replace existing hardcoded values with backend values, which are filtered by settings.
+            WI.CSSKeywordCompletions._propertyKeywordMap[property.name] = property.values;
```
are we 100% sure that this wont remove things that were previously exposed due to being hardcoded in `WI.CSSKeywordCompletions._propertyKeywordMap`?

also, this is not ideal for backwards compatibility (i.e. when inspecting an older iOS device that doesnt have the rest of the changes in this PR).  if my above concern is not an issue, then please change this to the following
```js
// COMPATIBILITY (macOS X.Y, iOS X.Y): the values from `CSS.getSupportedCSSProperties` were not yet exhaustive.
// Since support can't be tested directly, check for if the `Page.navigate` command has been removed.
// FIXME: Use explicit version checking once https://webkit.org/b/148680 is fixed.
if (InspectorBackend.hasCommand("Page.navigate"))
    WI.CSSKeywordCompletions.addPropertyCompletionValues(property.name, property.values);
else {
    // Replace existing hardcoded values with backend values, which are filtered by settings.
    WI.CSSKeywordCompletions._propertyKeywordMap[property.name] = property.values;
}
```

stwrt committed:
Web Inspector: Pull CSS property keyword completions from the backend instead of hardcoding them
https://bugs.webkit.org/show_bug.cgi?id=303627
rdar://problem/165914089

Reviewed by Tim Nguyen and Devin Rousso.

The Web Inspector's CSS autocomplete system previously hardcoded all valid
keyword values for CSS properties in CSSKeywordCompletions.js. This meant
that when new CSS values were added to WebCore, they had to be manually
duplicated in the Inspector frontend, and there was no way to filter values
based on runtime settings (like feature flags).

This change extends getSupportedCSSProperties() to return the valid keyword
values for each property, pulled directly from CSSProperties.json. The
frontend now uses these backend-provided values instead of the hardcoded
ones, ensuring completions stay in sync with WebCore and respect feature
flags.

Key changes:
- Added validKeywordsForProperty() to return keyword values from CSSProperties.json
- Added isKeywordValidForPropertyValues() to filter keywords by settings flags
- Added CSSParserContext(const Settings&) constructor for easy context creation
- Updated getSupportedCSSProperties() to include filtered values in the response
- Added settings-flag to grid-lanes/inline-grid-lanes values in CSSProperties.json

* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/scripts/process-css-properties.py:
(GenerateCSSPropertyNames._generate_css_property_names_gperf_prelude):
(GenerateCSSPropertyNames):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getSupportedCSSProperties):

Canonical link: https://commits.webkit.org/304195@main
